File: fd1d/animation/fd1d_1_4.py
#!/usr/bin/env python
# File: fd1d_1_4.py
# Name: D.Saravanan
# Date: 22/10/2021

""" Simulation of a propagating sinusoidal wave of 700 MHz
striking a medium with a relative dielectric constant of 4 """

import matplotlib.animation as animation
import matplotlib.pyplot as plt
import numpy as np

plt.style.use("classic")
plt.style.use("../pyplot.mplstyle")


def sinusoidal(t: int, ddx: float = 0.01, freq: float = 700e6) -> float:
	dt: float = ddx/6e8  # time step
	return np.sin(2 * np.pi * freq * dt * t)


def field(t: int, nx: int, cb: np.ndarray, ex: np.ndarray, hy: np.ndarray, bc: np.ndarray):
	# calculate the Hy field
	hy[0:nx-1] = hy[0:nx-1] + 0.5 * (ex[0:nx-1] - ex[1:nx])
	# calculate the Ex field
	ex[1:nx] = ex[1:nx] + cb[1:nx] * (hy[0:nx-1] - hy[1:nx])
	# put a sinusoidal wave at the low end
	ex[1] = ex[1] + sinusoidal(t, 0.01, 700e6)
	# absorbing boundary conditions
	ex[0], bc[0], bc[1] = bc[0], bc[1], ex[1]
	ex[nx-1], bc[3], bc[2] = bc[3], bc[2], ex[nx-2]


def dielectric(nx: int, epsr: float = 1) -> np.ndarray:
	cb = 0.5 * np.ones(nx, dtype=np.float64)
	cb[nx//2:] = 0.5/epsr
	return cb


def main():

	nx: int = 512
	ns: int = 3000

	ex = np.zeros(nx, dtype=np.float64)
	hy = np.zeros(nx, dtype=np.float64)

	bc = np.zeros(4, dtype=np.float64)

	epsr: float = 4 # relative permittivity
	cb: np.ndarray = dielectric(nx, epsr)

	# define the meta data for the movie
	fwriter = animation.writers["ffmpeg"]
	data = {"title": "Simulation of a sinusoidal hitting dielectric medium"}
	writer = fwriter(fps=15, metadata=data)

	# draw an empty plot, but preset the plot x- and y- limits
	fig, (ax1, ax2) = plt.subplots(2, sharex=False, gridspec_kw={"hspace": 0.2})
	fig.suptitle(r"FDTD simulation of a sinusoidal wave striking dielectric material")
	medium = (0.5/cb - 1)/(epsr - 1) if epsr > 1 else (0.5/cb - 1)
	medium[medium==0] = -1.5
	line1, = ax1.plot(ex, "k", lw=1)
	ax1.fill_between(range(nx), medium, medium[0], color='y', alpha=0.3)
	time_text = ax1.text(0.02, 0.90, "", transform=ax1.transAxes)
	epsr_txt1 = ax1.text(0.80, 0.80, "", transform=ax1.transAxes)
	ax1.set(xlim=(0, nx-1), ylim=(-1.2, 1.2), ylabel=r"$E_x$")
	ax1.set(xticks=range(0, nx+1, round(nx//10,-1)), yticks=np.arange(-1, 1.2, 1))
	line2, = ax2.plot(hy, "k", lw=1)
	ax2.fill_between(range(nx), medium, medium[0], color='y', alpha=0.3)
	epsr_txt2 = ax2.text(0.80, 0.80, "", transform=ax2.transAxes)
	ax2.set(xlim=(0, nx-1), ylim=(-1.2, 1.2), xlabel=r"FDTD cells", ylabel=r"$H_y$")
	ax2.set(xticks=range(0, nx+1, round(nx//10,-1)), yticks=np.arange(-1, 1.2, 1))

	with writer.saving(fig, "fd1d_1_4.mp4", 300):
		for t in range(1, ns+1):
			field(t, nx, cb, ex, hy, bc)

			line1.set_ydata(ex)
			time_text.set_text(f"T = {t}")
			epsr_txt1.set_text(f"epsr = {epsr}")
			line2.set_ydata(hy)
			epsr_txt2.set_text(f"epsr = {epsr}")
			writer.grab_frame()


if __name__ == "__main__":
	main()

File: fd1d/clang/fd1d_1_4.c
/* File: fd1d_1_4.c
 * Name: D.Saravanan
 * Date: 22/10/2021
 * Simulation of a propagating sinusoidal striking a dielectric medium
*/

#include <stdio.h>
#include <stdlib.h>
#include <math.h>


double sinusoidal(int t, double ds, double freq) {
    double dt = ds/6e8;  /* time step (s) */
    return sin(2*M_PI*freq*dt*t);
}


void exfield(int t, int nx, double *cb, double *ex, double *hy) {
    /* calculate the Ex field */
    for (int i = 1; i < nx; i++)
        ex[i] += cb[i] * (hy[i-1] - hy[i]);
    /* put a sinusoidal wave at the low end */
    ex[1] += sinusoidal(t, 0.01, 700e6);
}


void hyfield(int nx, double *ex, double *hy, double *bc) {
    /* absorbing boundary conditions */
    ex[0] = bc[0], bc[0] = bc[1], bc[1] = ex[1];
    ex[nx-1] = bc[3], bc[3] = bc[2], bc[2] = ex[nx-2];
    /* calculate the Hy field */
    for (int i = 0; i < nx-1; i++)
        hy[i] += 0.5 * (ex[i] - ex[i+1]);
}


double *dielectric(int nx, double epsr) {
    double *cb = (double*) calloc(nx, sizeof(*cb));
    for (int i = 0; i < nx; cb[i] = 0.5, i++);
    for (int i = nx/2; i < nx; cb[i] = 0.5/epsr, i++);
    return cb;
}


int main() {

    int nx = 512;  /* number of grid points */
    int ns = 740;  /* number of time steps */

    double *ex = (double*) calloc(nx, sizeof(*ex));
    double *hy = (double*) calloc(nx, sizeof(*hy));

    double bc[4] = {0.0};

    double ds = 0.01;  /* spatial step (m) */
    double dt = ds/6e8;  /* time step (s) */
    double epsr = 4.0;  /* relative permittivity */
    double *cb = dielectric(nx, epsr);

    for (int t = 1; t <= ns; t++) {
        exfield(t, nx, cb, ex, hy);
        hyfield(nx, ex, hy, bc);
    }

    free(cb);
    free(ex);
    free(hy);

    return 0;
}
